crud/src/main/java/com/redhat/lightblue/crud/TranslatorToJson.java

```diff
@@ -0,0 +1,203 @@
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.JsonNodeFactory;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.redhat.lightblue.metadata.ArrayElement;
+import com.redhat.lightblue.metadata.ArrayField;
+import com.redhat.lightblue.metadata.EntityMetadata;
+import com.redhat.lightblue.metadata.FieldCursor;
+import com.redhat.lightblue.metadata.FieldTreeNode;
+import com.redhat.lightblue.metadata.ObjectArrayElement;
+import com.redhat.lightblue.metadata.ObjectField;
+import com.redhat.lightblue.metadata.ReferenceField;
+import com.redhat.lightblue.metadata.SimpleArrayElement;
+import com.redhat.lightblue.metadata.SimpleField;
+import com.redhat.lightblue.metadata.Type;
+import com.redhat.lightblue.util.Error;
+import com.redhat.lightblue.util.JsonDoc;
+import com.redhat.lightblue.util.Path;
+
+/**
+ * Defines a class that take a response from a datasource and translates it
+ * into something Lightblue can use.
+ *
+ * @author dcrissman
+ *
+ * @param <S> - The source type that this {@link TranslatorToJson} converts
+ * to json.
+ */
+public abstract class TranslatorToJson<S> {
+
+    private final JsonNodeFactory factory;
+    protected final EntityMetadata entityMetadata;
+
+    public TranslatorToJson(JsonNodeFactory factory, EntityMetadata entityMetadata){
+        this.factory = factory;
+        this.entityMetadata = entityMetadata;
+    }
+
+    protected JsonNode toJson(Type type, Object value){
+        return type.toJson(factory, value);
+    }
+
+    /**
+     * Translates the source to a {@link JsonDoc}.
+     * @param source - Object containing the source data.
+     * @return {@link JsonDoc}
+     */
+    public JsonDoc translate(S source){
+        Error.push("translating to json");
+        try{
+            FieldCursor cursor = entityMetadata.getFieldCursor();
+
+            if (cursor.firstChild()) {
+                ObjectNode node = factory.objectNode();
+
+                iterateOverNodes(source, node, cursor);
+
+                return new JsonDoc(node);
+            }
+        }
+        finally{
+            Error.pop();
+        }
+
+        //TODO: What to do in case of a null value here?
+        return null;
+    }
+
+    private void iterateOverNodes(S source, ObjectNode targetNode, FieldCursor cursor){
+        do {
+            appendToJsonNode(source, targetNode, cursor);
+        } while(cursor.nextSibling());
+    }
+
+    void appendToJsonNode(S source, ObjectNode targetNode, FieldCursor fieldCursor){
+        targetNode.set(fieldCursor.getCurrentNode().getName(), translate(source, fieldCursor));
+    }
+
+    private JsonNode translate(S source, FieldCursor fieldCursor) {
+        FieldTreeNode field = fieldCursor.getCurrentNode();
+
+        Error.push(field.getFullPath().getLast());
+
+        try{
+            Object value = getValueFor(source, fieldCursor.getCurrentPath());
```

Comment:
It would require the implementing class to know what value to return from the source based on the path. I think a unit test for this would come in handy.